### PR TITLE
add `deactivate` support for tasks

### PR DESCRIPTION
**Pull request by andre-merzky**
This patch lets the bootstrapper create a `deactivate` script during
startup, which essentially mimics the venv `deactivate` command.  Alas,
the latter is not available in forked processes, so the executor cannot
deactivate the venv cleanly for the tasks.  With this script in place,
a task can add

    `. $RP_PILOT_SANDBOX/deactivate`

to the list of `pre_exec` commands to restore the original values for
`PATH`, `PYTHONPATH` and `PYTHONHOME`.  Any following `pre_exec` which
loads a venv or conda env should then not be affected by the pilot's
venv.  We might want to make this the default behavior at some point.

This patch also contains some other cleanups which got carried over by
cherry-picking the deactivate patch from an unrelated branch.

### CODE REVIEW

File: src/radical/pilot/unit_manager.py [hunk 1]
```diff
@@ -996,14 +996,19 @@ def wait_units(self, uids=None, state=None, timeout=None):
...
         with self._units_lock:
             states = [self._units[uid].state for uid in uids]
 
+        sdict = {state: states.count(state) for state in set(states)}
+        for state in sorted(set(states)):
+            self._rep.info('\t%-10s: %5d\n' % (state, sdict[state]))
+
+        if to_check: self._rep.warn('>>timeout\n')
+        else       : self._rep.ok  ('>>ok\n')
+
```
Comment (andre-merzky) [inline src/radical/pilot/unit_manager.py:1011]: Oh, another unrelated change which would have warranted it's own PR - sorry!  It triggers the reported to print a state summary after `wait_units()`.

File: src/radical/pilot/agent/bootstrap_0.sh [hunk 2]
```diff
@@ -146,6 +135,39 @@ export PYTHONNOUSERSITE=True
...
+                                        | cut -f 3 -d ' ' \
+                                        | cut -f 1 -d '=' \
+                                        | uniq)
+
+    # capture the original values in `deactivate`.  If no original value exists,
+    # unset the variable
+    for var in $changed
+    do
+        val=$(grep -e "^export $var=" env.orig | cut -f 2- -d '=')
+        if test -z "$val"
```
Comment (mtitov) [inline src/radical/pilot/agent/bootstrap_0.sh:161]: `if ! test -z "$val"` (currently it does opposite: unsets if original values exists, and export if not)

**Comment by mtitov** [comment]
In general, that's what I'd like to discuss - how to organize the environment for bootstrapping. I would propose to have some "isolated" (or "clean") environment, and only let for example `PATH`s & `PYTHON`s env variables pass through, everything else should be set by `pre_bootstrap_0/1` parameter(s) -> e.g. run the script with `env -i <cmd>` (should be researched more) and additionally to consider unalias processes (e.g., `unalias python`). And I would think that we wouldn't need to have to track changes in env-variables if we would have "isolated" environment.

p.s. related to these changes, if I'm not mistaken, `deactivate` file doesn't consider changes in env variables from `bootstrap_2.sh` (and specifically what will be set by `pre_bootstrap_1` -> `$PREBOOTSTRAP2_EXPANDED` internal variable).

p.p.s the usage of `deactivate` will be added later, right?

File: src/radical/pilot/agent/bootstrap_0.sh (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/bootstrap_0.sh:161]: Ah, thanks, nice catch

**Comment by andre-merzky** [comment]
We need to let more than `PATH` and `PYTHONPATH` through the bootstrapper, and specifically do not know in advance what we need to let through.  For example, the prte launch method needs the prte module loaded, which may need a specific compiler, but also sets other env variables like `PRTE_DIR`.  Similar for other launch methods, etc.  Those settings can vary from machine to machine, and those need to be visible to the different agent components.  Also, `PATH` and `PYTHONPATH` create most of the problems for task level virtualenvs, so the reset is needed for them anyway.

Ack on the `pre_bootstrap_0/1` settings.  They will be caught in many cases (those are usually a subset of `pre_bootstrap_0`), but that should be made more explicit.

As for usage: at the moment we can add the deactivate call manually to the task description, to experiment how it behaves, specifically in cases where the application breaks otherwise.  An automated use makes sense once we see that this is stable and has no side effects.

File: src/radical/pilot/agent/bootstrap_0.sh (hunk 2 above)
Comment (andre-merzky) [reply src/radical/pilot/agent/bootstrap_0.sh]: I realized that this wasn't correct anyway: an empty value does not mean that the var did not exist.  This is thus fixed differently now - can you have a look?

**Comment by andre-merzky** [comment]
Yeah, I am also not happy, yet.  Thanks for checking.

File: src/radical/pilot/agent/bootstrap_0.sh [hunk 3]
```diff
@@ -157,12 +157,14 @@ create_deactivate()
     # unset the variable
     for var in $changed
     do
-        val=$(grep -e "^export $var=" env.orig | cut -f 2- -d '=')
-        if test -z "$val"
+        orig=$(grep -e "^export $var=")
```
Comment (mtitov) [inline src/radical/pilot/agent/bootstrap_0.sh:160]: string was overcut.. missed the variable `env.org` -> `orig=$(grep -e "^export $var=" env.orig)`

Comment (andre-merzky) [reply src/radical/pilot/agent/bootstrap_0.sh]: Yeah, saw that too - I pushed before the test run completed... *blush*

**Comment by mtitov** [comment]
@andre-merzky ack the code correctness

BTW as a follow up question: env variable `$RP_PILOT_SANDBOX` is set/managed somewhere? cannot check it..

**Comment by andre-merzky** [comment]
`RP_PILOT_SANDBOX`: yes, it is added [here](https://github.com/radical-cybertools/radical.pilot/pull/2127/files#diff-a629ccc4c2ef78f37f7e121382ce2677R213), which adds it to the tasks execution shell script,
